buildAndReleaseTask/serviceEndpoint.ts

```diff
@@ -47,4 +56,84 @@ export function getServiceEndpoint(
 
     return endpoint;
 }
+
+/**
+ * Tries to get the endpoint details for a workload identity federation
+ * service connection as per V2 of the Azure CLI task
+ * https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/AzureCLIV2/azureclitask.ts#L143-L152
+ */
+async function tryGetOidcServiceEndpoint(
+    connectedServiceName: string
+): Promise<IServiceEndpoint | undefined> {
+    const authScheme = tl.getEndpointAuthorizationScheme(
+        connectedServiceName,
+        true
+    );
+    if (authScheme?.toLowerCase() !== "workloadidentityfederation") {
+        return undefined;
+    }
+
+    const federatedToken = await tryGetIdToken(connectedServiceName);
+
+    if (!federatedToken) {
+        return undefined;
+    }
+
+    tl.setSecret(federatedToken);
+
+    const endpoint = {
+        clientId: tl.getEndpointAuthorizationParameter(
+            connectedServiceName,
+            "serviceprincipalid",
+            false
+        ),
+        oidcToken: await tryGetIdToken(connectedServiceName),
+        // It is entirely possible subscriptionId is not present, so setting subscription as optional stops an unintended failure.
+        // eg- a ManagementGroup scoped SP has access to multiple subscriptions; the subscription that is the target of a pulumi up will be enforced in the pulumi program config/env.
+        subscriptionId: tl.getEndpointDataParameter(
+            connectedServiceName,
+            "subscriptionid",
+            true
+        ),
+        tenantId: tl.getEndpointAuthorizationParameter(
+            connectedServiceName,
+            "tenantid",
+            false
+        ),
+    } as IServiceEndpoint;
+
+    return endpoint;
+}
+
+/**
+ * Tries to get the ID token as per V2 of the Azure CLI task
+ * https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/AzureCLIV2/azureclitask.ts#L251-L268
+ */
+async function tryGetIdToken(
+    connectedService: string
+): Promise<string | undefined | null> {
+    const jobId = tl.getVariable("System.JobId")!;
+    const planId = tl.getVariable("System.PlanId")!;
+    const projectId = tl.getVariable("System.TeamProjectId")!;
+    const hub = tl.getVariable("System.HostType")!;
+    const uri = tl.getVariable("System.CollectionUri")!;
+    const token = getSystemAccessToken();
+
+    const authHandler = getHandlerFromToken(token);
+    const connection = new WebApi(uri, authHandler);
+    const api: ITaskApi = await connection.getTaskApi();
+    const response = await api.createOidcToken(
```

Comment:
What's the lifetime of this token? As far as I can see, there's no automatic renewal, so could pipeline jobs fail part way through when it expires?

Reply:
My understanding is that the OIDC token will exchanged with Azure by the provider.
I think it happens somewhere in [provider/pkg/provider/auth.go](https://github.com/pulumi/pulumi-azure-native/blob/f7ffc17a30cf5b602603034620f6fe1b24cf08d1/provider/pkg/provider/auth.go) or in [Azure/go-autorest](https://github.com/Azure/go-autorest).